### DISCUSSION TITLE
Polymorphism over python classes in rust

### DISCUSSION THREAD

**Discussion by jayphur** [PyO3/pyo3#3453] [Questions] [answered]
Hello, 

I was wondering: Is it possible to: 

Define a parent in class in my rust api: 
 ```rust
#[pyclass(subclass)
pub struct ParentClass{
      // stuff
}
#[pymethods]
impl ParentClass{
      fn do_stuff(&self) { 
             // default implementation
       }
      fn get_two(&self) -> u8{ 
             2
       }
}
```
Then, have python define children of this class:
```python
from rust_api import ParentClass
class Child(ParentClass):
        def do_stuff(self):
                # python version
```
...and have this new child class be usable at runtime by rust code (possibly loaded from something like `PyModule::from_code()`).

I've seen something like this in the user guide with both the parent class/struct and child class/struct defined in rust.  

The implementation of a system like (how it relates to the GIL, lifetimes, safety, etc) this does not matter to me. I am just wondering if something like this was possible.

Thanks :)

**Comment by birkenfeld** [comment]
You can do that, you just need to call methods via the Python API, not directly on the Rust level.

**Reply by jayphur** [reply]
Do you mean with something like `call_method(...)`?

**Reply by davidhewitt** [reply, marked as answer]
Yep exactly, see for example the older discussion https://github.com/PyO3/pyo3/issues/1875#issuecomment-922367642
